### PR TITLE
CDAP-3472 Correct error messsage when trying to stop the MapReduce pr…

### PR DESCRIPTION
**Pull request by sagarkapare**
…ogram started by Workflow.

JIRA: https://issues.cask.co/browse/CDAP-3472

### CODE REVIEW

**Comment by sagarkapare** [comment]
Release build is running here - http://builds.cask.co/browse/CDAP-RBT430

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 1]
```diff
@@ -1354,11 +1354,16 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
...
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord.getProperties().containsKey("workflowrunid")
+          && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+          throw new BadRequestException("MapReduce program run started by Workflow cannot be stopped. " +
+                                          "Please stop the Workflow.");
+        }
       }
+      throw new NotFoundException(new Id.Run(identifier, runId));
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: runId might be null in this case, right?

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Also, you're not properly responded with BadRequestException("Program not running") like before. Isn't that still a possibility?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 2]
```diff
@@ -1354,11 +1354,16 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
...
-      } else if (runId == null) {
-        throw new BadRequestException("Program not running");
-      } else {
-        throw new NotFoundException(new Id.Run(identifier, runId));
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord.getProperties().containsKey("workflowrunid")
+          && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+          throw new BadRequestException("MapReduce program run started by Workflow cannot be stopped. " +
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: should we add the mapreduce and workflow ids in the message?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 3]
```diff
@@ -1354,11 +1354,16 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
...
       } else if (!store.programExists(identifier)) {
         throw new ProgramNotFoundException(identifier);
-      } else if (runId == null) {
-        throw new BadRequestException("Program not running");
-      } else {
-        throw new NotFoundException(new Id.Run(identifier, runId));
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord.getProperties().containsKey("workflowrunid")
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: what is the expected behavior if this condition is false?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: yeah something seems to be missing here. perhaps this should move one line above. and there should be a ``BadRequestException(String.format("Program %s not running")`` on this line. I'm not entirely sure, though.

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Fixed the code to throw BadRequestException("Program not running") when the run id is null.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: It should actually throw NotFoundException when the condition is false. Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I think currently its not possible to get the program name from runid.

**Comment by sagarkapare** [comment]
@bdmogal Addressed comments. Please take a look again.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 2 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: We have access to a variable named 'identifier' which contains this information.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 4]
```diff
@@ -1354,11 +1354,17 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
...
         throw new ProgramNotFoundException(identifier);
-      } else if (runId == null) {
-        throw new BadRequestException("Program not running");
-      } else {
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
+          && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+          throw new BadRequestException("MapReduce program run started by Workflow cannot be stopped. " +
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: lets add atleast the mapreduce name and run id, along with the workflow runid as well.

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Fixed error message.

**Comment by bdmogal** [comment]
LGTM. Merge after build succeeds.

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-RBT430-3

**Comment by sagarkapare** [comment]
Build passed here - http://builds.cask.co/browse/CDAP-RBT430-6. Merging.

**Pull request merged**